library/ssl_msg.c

```diff
@@ -1596,6 +1596,21 @@ int mbedtls_ssl_decrypt_buf(mbedtls_ssl_context const *ssl,
         defined(MBEDTLS_SSL_PROTO_TLS1_2)
         if (transform->minor_ver > MBEDTLS_SSL_MINOR_VERSION_0) {
             /*
+             * Delicate code warning: Lucky Thirteen countermeasure.
+             *
+             * In CBC+HMAC cipher suites, the MAC is near the end of the
+             * ciphertext, followed with 1 to 256 bytes. To avoid a
+             * padding oracle attack known as Lucky Thirteen, we must
+             * not reveal the length of the padding, and therefore the
+             * position of the MAC, through timing. This includes the
+             * timing of memory accesses, which can leak to a local
+             * attacker. Hence we use the function mbedtls_ct_hmac()
+             * which is specially crafted to verify a HMAC at an
+             * unknown position within a range.
```

Comment:
I don't think that's what `mbedtls_ct_hmac()` does: it does not verify the mac but only computes it. More precisely, it computes the HMAC of a message of secret length withing a range without leaking that length.

Comparison with the peer's value without leaking its position comes a few lines later: `mbedtls_ct_memcpy_offset()` copies from a secret position withing a range to a local buffer, then we can finally do a simple `mbedtls_ct_memcmp()`.

(Note: I think "secret" is more accurate than "unknown" here.)

Reply:
I force-pushed a new version where I improved the wording in various places (comments in code, second commit message, changelog).